### PR TITLE
Port MetalKernels

### PR DESCRIPTION
**Pull request by maxwindiff**
## Outstanding issues

- [x] `copyto_testsuite` used to fail for strange reasons, but seems consistently passing now?

## Unsupported currently

- [x] `get_backend(sparse)`: Depends on `sparse(MtlArray)`, skipped
- [x] `convert`: Depends on #69, skipped
- [x] `specialfunction`: Missing [device intrinsics](https://github.com/JuliaGPU/CUDA.jl/blob/master/src/device/intrinsics/special_math.jl) in Metal, skipped

### CODE REVIEW

**Comment by maleadt** [comment]
cc @vchuravy @tgymnich

File: src/Metal.jl [hunk 1]
```diff
@@ -50,4 +50,8 @@ include("gpuarrays.jl")
 include("../lib/mps/MPS.jl")
 export MPS
 
+# KernelAbstractions
+include("MetalKernels.jl")
+export CUDABackend
```
Comment (tgymnich) [inline src/Metal.jl:55]: CUDA?

**Comment by maleadt** [comment]
Can we make use of the specialized `thread_position_in_threadgroup_1d` and `thread_position_in_grid_1d` queries instead of having to multiply things (i.e., https://github.com/JuliaGPU/Metal.jl/pull/76 and https://github.com/JuliaGPU/Metal.jl/pull/139)?

**Comment by maxwindiff** [comment]
I think the original version by @tgymnich already uses those queries - https://github.com/JuliaGPU/Metal.jl/pull/131/files#diff-2c3eb99b21164848c74517fc0b0e49e5d8785051cd61bf8ecda6fb5e2dfcfb22R146-R168

**Comment by maleadt** [comment]
Ah, right, not sure how I overlooked those.

**Comment by vchuravy** [comment]
We should expose a query in KA that checks capabilities of the backend. E.g. atomics and `Float64` support come to mind.

File: src/MetalKernels.jl [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+    code = quote
+        $GPUCompiler.@override($Metal.method_table, $ex)
+    end
+    if isdefined(Base.Experimental, Symbol("@overlay"))
+        return esc(code)
+    else
+        push!(overrides, code)
+        return
+    end
+end
```
Comment (vchuravy) [inline src/MetalKernels.jl:129]: Metal.jl doesn't support Julia 1.6 so you can just directly use `Metal.@device_override`

File: src/MetalKernels.jl [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+
+function __init__()
+    precompiling = ccall(:jl_generating_output, Cint, ()) != 0
+    precompiling && return
+    # register device overrides
+    eval(Expr(:block, overrides...))
+    empty!(overrides)
+end
```
Comment (vchuravy) [inline src/MetalKernels.jl:137]:
```suggestion
```

File: src/MetalKernels.jl (hunk 2 above)
Comment (maxwindiff) [reply src/MetalKernels.jl]: Done, thanks for the tip.

File: src/MetalKernels.jl (hunk 3 above)
Comment (maxwindiff) [reply src/MetalKernels.jl]: Done

**Comment by maxwindiff** [comment]
@vchuravy Sounds good, I'll take a stab at adding it.

By the way I have a question about `specialfunctions`. The test fails with this error:

```julia
julia> a = MtlArray(Float32[1.0])
1-element MtlVector{Float32}:
 1.0

julia> gamma.(a)
ERROR: InvalidIRError: compiling broadcast_kernel(Metal.mtlKernelContext, MtlDeviceVector{Float32, 1}, Val{CartesianIndices((1,))}, Base.Broadcast.Broadcasted{Metal.MtlArrayStyle{1}, Tuple{Base.OneTo{Int64}}, typeof(gamma), Tuple{Base.Broadcast.Extruded{MtlDeviceVector{Float32, 1}, Tuple{Bool}, Tuple{Int64}}}}, Int64) in world 32495 resulted in invalid LLVM IR
Reason: unsupported call to an unknown function (call to gpu_malloc)
...
Reason: unsupported call through a literal pointer (call to tgammaf)
Stacktrace:
 [1] _gamma
   @ ~/.julia/packages/SpecialFunctions/gXPNz/src/gamma.jl:578
 [2] gamma
   @ ~/.julia/packages/SpecialFunctions/gXPNz/src/gamma.jl:567
 [3] _broadcast_getindex_evalf
   @ ./broadcast.jl:670
...
```

The second error complains about a [`ccall((:tgammaf, libopenlibm), ...)`](https://github.com/JuliaMath/SpecialFunctions.jl/blob/master/src/gamma.jl#L578). How did that work for the other backends?

**Comment by maxwindiff** [comment]
Oh found it - https://github.com/JuliaGPU/CUDA.jl/blob/master/src/device/intrinsics/special_math.jl
There doesn't seem to be an equivalent function in Metal however.

File: src/MetalKernels.jl [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+KernelAbstractions.ones(::MetalBackend, ::Type{T}, dims::Tuple) where T = Metal.ones(T, dims)
+
+# Import through parent
+import KernelAbstractions: StaticArrays, Adapt
+import .StaticArrays: MArray
+
+KernelAbstractions.get_backend(::Metal.MtlArray) = MetalBackend()
+KernelAbstractions.synchronize(::MetalBackend) = Metal.synchronize()
+
+# TODO: why are these not needed in https://github.com/JuliaGPU/CUDA.jl/pull/1772 ?
```
Comment (maxwindiff) [inline src/MetalKernels.jl:24]: @vchuravy Do you know why these aren't needed in https://github.com/JuliaGPU/CUDA.jl/pull/1772?

Comment (maxwindiff) [reply src/MetalKernels.jl:24]: Umm.. however I don't see them on the [`release-0.8`](https://github.com/JuliaGPU/KernelAbstractions.jl/blob/release-0.8/lib/CUDAKernels/src/CUDAKernels.jl) branch either.

**Comment by maxwindiff** [comment]
Once https://github.com/JuliaGPU/KernelAbstractions.jl/pull/369 and https://github.com/JuliaGPU/KernelAbstractions.jl/pull/374 are merged, most tests should pass. There's one weird issue remaining however -- the `copyto` tests fails *if* the `adapt` unit test was run before it. Still trying to figure out why...

To disable unrelated tests:

```diff
diff --git a/test/kernelabstractions.jl b/test/kernelabstractions.jl
index 6542799..17c79ff 100644
--- a/test/kernelabstractions.jl
+++ b/test/kernelabstractions.jl
@@ -13,4 +13,8 @@ Testsuite.testsuite(()->MetalBackend(), "Metal", Metal, MtlArray, Metal.MtlDevic
     "Convert",           # depends on https://github.com/JuliaGPU/Metal.jl/issues/69
     "SpecialFunctions",  # no equivalent Metal intrinsics for gamma, erf, etc
     "sparse",            # not supported yet
+
+    "partition", "get_backend", "indextest", "Const", "CPU synchronization",
+    "Zero iteration space $(MetalBackend())", "return statement", "fallback test: callable types", "priority",
+    "Localmem", "Private", "Unroll", "Printing", "Compiler", "Reflection", "Examples",
 ]))
```

File: src/MetalKernels.jl (hunk 4 above)
Comment (vchuravy) [reply src/MetalKernels.jl:24]: Yeah that's a new feature

**Comment by vchuravy** [comment]
>  There doesn't seem to be an equivalent function in Metal however.

Yeah we just need to skip that test.

**Comment by maxwindiff** [comment]
I didn't change anything except updating dependencies, but now the `copyto` failure is gone. I guess this is ready for review. @vchuravy @tgymnich Appreciate if you can take another look, esp for the `Adapt` and `copyto` implementations.

**Review by vchuravy — APPROVED** [review]

**Review by tgymnich — APPROVED** [review]

File: test/kernelabstractions.jl [hunk 5]
```diff
@@ -0,0 +1,16 @@
+import KernelAbstractions
```
Comment (maleadt) [inline test/kernelabstractions.jl:1]: This test file is included by Metal.jl's test runner, so shouldn't have to import Metal or Test. Also, is importing MetalKernels required? The module is loaded already, and you're not relying on functionality from it directly (only from KA.jl's functionality).

File: test/kernelabstractions.jl [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+using Test
+
+include(joinpath(dirname(pathof(KernelAbstractions)), "..", "test", "testsuite.jl"))
+
+using Metal
+using Metal.MetalKernels
+
+# TODO: check if Metal is functional (depends on https://github.com/JuliaGPU/Metal.jl/issues/121)
+Metal.versioninfo()
+Metal.allowscalar(false)
```
Comment (maleadt) [inline test/kernelabstractions.jl:11]:
```suggestion
```

File: test/kernelabstractions.jl [hunk 7]
```diff
@@ -0,0 +1,16 @@
+import KernelAbstractions
+using Test
```
Comment (maleadt) [inline test/kernelabstractions.jl:2]:
```suggestion
```

File: test/kernelabstractions.jl (hunk 6 above)
Comment (maleadt) [inline test/kernelabstractions.jl:11]:
```suggestion
```

File: test/kernelabstractions.jl (hunk 5 above)
Comment (maxwindiff) [reply test/kernelabstractions.jl:1]: You are right, most of these are not needed. The test does use `MetalKernel.MetalBackend()` so I added `MetaKernels` back, but let me know if it should be changed to a more narrow import.

File: Manifest.toml [hunk 8]
```diff
@@ -75,6 +81,14 @@ git-tree-sha1 = "abc9885a7ca2052a736a600f7fa66209f96506e1"
 uuid = "692b3bcd-3c85-4b1f-b108-f13ce0eb3210"
 version = "1.4.1"
 
+[[deps.KernelAbstractions]]
+deps = ["Adapt", "Atomix", "InteractiveUtils", "LinearAlgebra", "MacroTools", "SnoopPrecompile", "SparseArrays", "StaticArrays", "UUIDs", "UnsafeAtomics", "UnsafeAtomicsLLVM"]
+git-tree-sha1 = "1f6564e25c2e6e33f1f5caaf4b272d48e3c24a9f"
+repo-rev = "main"
```
Comment (maxwindiff) [inline Manifest.toml:87]: Need to be updated once 0.9.1 is released.

**Review by vchuravy — APPROVED** [review]

File: src/Metal.jl [hunk 9]
```diff
@@ -50,4 +50,8 @@ include("gpuarrays.jl")
 include("../lib/mps/MPS.jl")
 export MPS
 
+# KernelAbstractions
+include("MetalKernels.jl")
```
Comment (vchuravy) [inline src/Metal.jl:54]:
```suggestion
include("MetalKernels.jl")
import .MetalKernels: MetalBackend
```

**Pull request merged**